src/leap/mail/mail.py

```diff
@@ -0,0 +1,697 @@
+                         self.mbox_uuid, h))
+            return defer.gatherResults(d), hashes
+
+        def delete_uid_entries((uids, hashes)):
+            d = []
+            for h in hashes:
+                d.append(self.mbox_indexer.delete_doc_by_hash(
+                         self.mbox_uuid, h))
+            return defer.gatherResults(d).addCallback(
+                lambda _: uids)
+
+        mdocs_deleted = self.adaptor.del_all_flagged_messages(
+            self.store, self.mbox_uuid)
+        mdocs_deleted.addCallback(get_uid_list)
+        mdocs_deleted.addCallback(delete_uid_entries)
+        return mdocs_deleted
+
+    # TODO should add a delete-by-uid to collection?
+
+    def delete_all_docs(self):
+        def del_all_uid(uid_list):
+            deferreds = []
+            for uid in uid_list:
+                d = self.get_message_by_uid(uid)
+                d.addCallback(lambda msg: msg.delete())
+                deferreds.append(d)
+            return defer.gatherResults(deferreds)
+
+        d = self.all_uid_iter()
+        d.addCallback(del_all_uid)
+        return d
+
+    def _update_flags_or_tags(self, old, new, mode):
+        if mode == Flagsmode.APPEND:
+            final = list((set(tuple(old) + new)))
+        elif mode == Flagsmode.REMOVE:
+            final = list(set(old).difference(set(new)))
+        elif mode == Flagsmode.SET:
+            final = new
+        return final
+
+    def udpate_flags(self, msg, flags, mode):
+        """
+        Update flags for a given message.
+        """
+        wrapper = msg.get_wrapper()
+        current = wrapper.fdoc.flags
+        newflags = self._update_flags_or_tags(current, flags, mode)
+        wrapper.fdoc.flags = newflags
+
+        wrapper.fdoc.seen = MessageFlags.SEEN_FLAG in newflags
+        wrapper.fdoc.deleted = MessageFlags.DELETED_FLAG in newflags
+
+        return self.adaptor.update_msg(self.store, msg)
+
+    def update_tags(self, msg, tags, mode):
+        """
+        Update tags for a given message.
+        """
+        wrapper = msg.get_wrapper()
+        current = wrapper.fdoc.tags
+        newtags = self._update_flags_or_tags(current, tags, mode)
+        wrapper.fdoc.tags = newtags
+        return self.adaptor.update_msg(self.store, msg)
+
+
+class Account(object):
+    """
+    Account is the top level abstraction to access collections of messages
+    associated with a LEAP Mail Account.
+
+    It primarily handles creation and access of Mailboxes, which will be the
+    basic collection handled by traditional MUAs, but it can also handle other
+    types of Collections (tag based, for instance).
+
+    leap.mail.imap.SoledadBackedAccount partially proxies methods in this
+    class.
+    """
+
+    # Adaptor is passed to the returned MessageCollections, so if you want to
+    # use a different adaptor this is the place to change it, by subclassing
+    # the Account class.
+
+    adaptor_class = SoledadMailAdaptor
+    store = None
```

Comment:
Is there any need for the `store = None` as init will rewrite it? The only case I can come up is for classmethods, but there is none in this class.